main/uart0_driver.c

```diff
@@ -0,0 +1,178 @@
+    }
+    return data;
+}
+
+static FORCE_INLINE uint8_t *uart0_tx_fifo(uint8_t* data, size_t size) {
+    for (const uint8_t* end = data + size; data != end; ++data) {
+        uart0.fifo.rw_byte = *data;
+    }
+    return data;
+}
+
+static void IRAM_ATTR uart0_isr(void *param)
+{
+    BaseType_t task_woken = pdFALSE;
+    for (;;) {
+        const uint32_t uart_intr_status = uart0.int_st.val;
+
+        if (uart_intr_status == 0) {
+            if (task_woken == pdTRUE) {
+                portYIELD_FROM_ISR();
+            }
+            return;
+        }
+
+        if (uart_intr_status & UART_TXFIFO_EMPTY_INT_ST_M) {
+            const uint8_t tx_remain_fifo_cnt = UART_FIFO_LEN - uart0.status.txfifo_cnt;
+            if (tx_size > tx_remain_fifo_cnt) {
+                // send partial data to fifo and keep interrupt enabled
+                tx_data = uart0_tx_fifo(tx_data, tx_remain_fifo_cnt);
+                tx_size -= tx_remain_fifo_cnt;
+            } else {
+                // send complete data to fifo
+                (void)uart0_tx_fifo(tx_data, tx_size);
+
+                // disable more interrupts and unblock uart0_tx_bytes()
+                uart0.int_ena.txfifo_empty = 0;
+                vTaskNotifyGiveFromISR(tx_task_handle, &task_woken);
+            }
+            uart0.int_clr.txfifo_empty = 1;
+        } else if (uart_intr_status & (UART_RXFIFO_FULL_INT_ST_M|UART_RXFIFO_TOUT_INT_ST_M)) {
+            const uint8_t rx_fifo_cnt = uart0.status.rxfifo_cnt;
+            if (rx_size > rx_fifo_cnt) {
+                // load partial data from fifo and keep interrupt enabled
+                rx_data = uart0_rx_fifo(rx_data, rx_fifo_cnt);
+                rx_size -= rx_fifo_cnt;
+            } else {
+                // load complete data from fifo
+                (void)uart0_rx_fifo(rx_data, rx_size);
+
+                // disable more interrupts and unblock uart0_rx_bytes()
+                uart0.int_ena.rxfifo_full = 0;
+                uart0.int_ena.rxfifo_tout = 0;
+                vTaskNotifyGiveFromISR(rx_task_handle, &task_woken);
+            }
+            uart0.int_clr.rxfifo_full = 1;
+            uart0.int_clr.rxfifo_tout = 1;
+        } else if (uart_intr_status & UART_RXFIFO_OVF_INT_ST_M) {
+            // When fifo overflows, we reset the fifo.
+            uart0.conf0.rxfifo_rst = 0x1;
+            uart0.conf0.rxfifo_rst = 0x0;
+            uart0.int_clr.rxfifo_ovf = 1;
+        } else {
+            // TODO: Think more about overflows, frame errors and such
+            uart0.int_clr.val = uart_intr_status; // simply clear all other intr status
+        }
+    }
+}
+
+void IRAM_ATTR uart0_tx_bytes(uint8_t *data, size_t size) {
+    const uint8_t tx_remain_fifo_cnt = UART_FIFO_LEN - uart0.status.txfifo_cnt;
+    if (size > tx_remain_fifo_cnt) {
+        // send partial data to fifo
+        tx_data = uart0_tx_fifo(data, tx_remain_fifo_cnt);
+        tx_size = size - tx_remain_fifo_cnt;
+
```

Comment:
I understand. Again, baudrate + typical time to fill/drain fifo are such that I never observed this, but I guess I can extend the critical section there.

Reply:
Thinking about it now, I do not understand what I understood yesterday. :upside_down_face: 

What I think would happen is that UART hardware will just empty the fifo and then we will not fill it entirely, but it does not matter, because we will then still continue in the interrupt handler and send the rest of the stuff.

Basically `tx_remain_fifo_cnt` is there only to ensure we do not write more than the FIFO capacity. If we write less, we will catch up later anyway. Unless we are already sending the entire thing and we only do that when, again, there is enough space in the FIFO.

Does that make sense?

Reply:
I have adjusted the scenario:
1. Remaining fifo count is 96 (after next transmitted byte the empty interrupt will be generated 128-32=96)
2. Data size to be send is > 97
3. Part of data is added to the queue
4. If we are unlucky now the queue is empty and interrupt is generated(but not enabled yet)
5. Interrupt is enabled
6. Wait for interrupt to send the second part of the data

The point is that the (4) never happens as either the data fits completely in the queue or the queue is filled and the empty interrupt will come later. Let's hope this will hold every time.

Reply:
As I understand it, there are two different registers for state. One contains raw interrupt state, second is latched based on interrupt enabled register. We are using latched register in ISR, not raw. Which I think prevents all the problems you mentioned.